sweepbatcher/sweep_batcher.go

```diff
@@ -0,0 +1,654 @@
+
+	// quit signals that the batch must stop.
+	quit chan struct{}
+
+	// wallet is the wallet kit client that is used by batches.
+	wallet lndclient.WalletKitClient
+
+	// chainNotifier is the chain notifier client that is used by batches.
+	chainNotifier lndclient.ChainNotifierClient
+
+	// signerClient is the signer client that is used by batches.
+	signerClient lndclient.SignerClient
+
+	// musig2ServerKit includes all the required functionality to collect
+	// and verify signatures by the swap server in order to cooperatively
+	// sweep funds.
+	musig2ServerSign MuSig2SignSweep
+
+	// verifySchnorrSig is a function that can be used to verify a schnorr
+	// signature.
+	VerifySchnorrSig VerifySchnorrSig
+
+	// chainParams are the chain parameters of the chain that is used by
+	// batches.
+	chainParams *chaincfg.Params
+
+	// store includes all the database interactions that are needed by the
+	// batcher and the batches.
+	store BatcherStore
+
+	// swapStore includes all the database interactions that are needed for
+	// interacting with swaps.
+	swapStore loopdb.SwapStore
+
+	// wg is a waitgroup that is used to wait for all the goroutines to
+	// exit.
+	wg sync.WaitGroup
+}
+
+// NewBatcher creates a new Batcher instance.
+func NewBatcher(wallet lndclient.WalletKitClient,
+	chainNotifier lndclient.ChainNotifierClient,
+	signerClient lndclient.SignerClient, musig2ServerSigner MuSig2SignSweep,
+	verifySchnorrSig VerifySchnorrSig, chainparams *chaincfg.Params,
+	store BatcherStore, swapStore loopdb.SwapStore) *Batcher {
+
+	return &Batcher{
+		batches:          make(map[int32]*batch),
+		sweepReqs:        make(chan SweepRequest),
+		errChan:          make(chan error, 1),
+		completeChan:     make(chan int32),
+		quit:             make(chan struct{}),
+		wallet:           wallet,
+		chainNotifier:    chainNotifier,
+		signerClient:     signerClient,
+		musig2ServerSign: musig2ServerSigner,
+		VerifySchnorrSig: verifySchnorrSig,
+		chainParams:      chainparams,
+		store:            store,
+		swapStore:        swapStore,
+	}
+}
+
+// Run starts the batcher and processes incoming sweep requests.
+func (b *Batcher) Run(ctx context.Context) error {
+	runCtx, cancel := context.WithCancel(ctx)
+	defer func() {
+		cancel()
+
+		for _, batch := range b.batches {
+			batch.Wait()
+		}
+
+		b.wg.Wait()
+	}()
+
+	// First we fetch all the batches that are not in a confirmed state from
+	// the database. We will then resume the execution of these batches.
+	batches, err := b.FetchUnconfirmedBatches(runCtx)
+	if err != nil {
+		return err
+	}
+
+	for _, batch := range batches {
+		err := b.spinUpBatchFromDB(runCtx, batch)
+		if err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case sweepReq := <-b.sweepReqs:
+			sweep, err := b.fetchSweep(runCtx, sweepReq)
+			if err != nil {
+				return err
+			}
+
+			err = b.handleSweep(runCtx, sweep, sweepReq.Notifier)
+			if err != nil {
+				return err
+			}
+
+		case id := <-b.completeChan:
```

Comment:
I think instead of using the `completeChan` we should lazily delete the completed batches when we add the next sweep (in `handleSweep`). Needs an `isComplete()` call on the `batch`. This would avoid the potential deadlock when the batcher quits, and waits for the batches in the deferred function above while there's at least on who wants to write to the `completeChan`.

Reply:
Could instead avoid the deadlock by blocking on both `completeChan` and `ctx.Done` on the batch's side, this way this edge-case would simply exit instead of deadlocking
Wdyt?

Reply:
Ok that would lead to a new potential deadlock. Just gonna convert to lazy-delete